From c80ccf84e9024c0d3d5388908aa04ecd81d223ee Mon Sep 17 00:00:00 2001
From: Kamil Dudka <kdudka@redhat.com>
Date: Thu, 21 Jul 2016 16:26:34 +0200
Subject: [PATCH] tests/secureserver.pl: work around RHBZ #1358810

Bug: https://bugzilla.redhat.com/1358810
---
 tests/secureserver.pl | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tests/secureserver.pl b/tests/secureserver.pl
index 3a7443c..c2da40d 100755
--- a/tests/secureserver.pl
+++ b/tests/secureserver.pl
@@ -275,6 +275,7 @@ if($stunnel_version >= 400) {
         print STUNCONF "cert = $certfile\n";
         print STUNCONF "debug = $loglevel\n";
         print STUNCONF "socket = $socketopt\n";
+        print STUNCONF "requireCert = no\n";
         if($fips_support) {
             # disable fips in case OpenSSL doesn't support it
             print STUNCONF "fips = no\n";
-- 
2.5.5

