 configure.ac |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/configure.ac b/configure.ac
index af30b8a..ca6c7cc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -275,7 +275,10 @@ dnl **********************************************************************
 
 CURL_CHECK_COMPILER
 CURL_SET_COMPILER_BASIC_OPTS
-CURL_SET_COMPILER_DEBUG_OPTS
+
+dnl do not perturb CFLAGS given by the build system
+dnl CURL_SET_COMPILER_DEBUG_OPTS
+
 CURL_SET_COMPILER_OPTIMIZE_OPTS
 CURL_SET_COMPILER_WARNING_OPTS
 
--- a/configure
+++ b/configure
@@ -15027,190 +15026,6 @@
   fi
 
 
-        #
-  if test "$compiler_id" != "unknown"; then
-    #
-    tmp_save_CFLAGS="$CFLAGS"
-    tmp_save_CPPFLAGS="$CPPFLAGS"
-    #
-    tmp_options=""
-    tmp_CFLAGS="$CFLAGS"
-    tmp_CPPFLAGS="$CPPFLAGS"
-
-    ac_var_stripped=""
-  for word1 in $tmp_CFLAGS; do
-    ac_var_strip_word="no"
-    for word2 in $flags_dbg_all; do
-      if test "$word1" = "$word2"; then
-        ac_var_strip_word="yes"
-      fi
-    done
-    if test "$ac_var_strip_word" = "no"; then
-      ac_var_stripped="$ac_var_stripped $word1"
-    fi
-  done
-    tmp_CFLAGS="$ac_var_stripped"
-  squeeze tmp_CFLAGS
-
-
-    ac_var_stripped=""
-  for word1 in $tmp_CPPFLAGS; do
-    ac_var_strip_word="no"
-    for word2 in $flags_dbg_all; do
-      if test "$word1" = "$word2"; then
-        ac_var_strip_word="yes"
-      fi
-    done
-    if test "$ac_var_strip_word" = "no"; then
-      ac_var_stripped="$ac_var_stripped $word1"
-    fi
-  done
-    tmp_CPPFLAGS="$ac_var_stripped"
-  squeeze tmp_CPPFLAGS
-
-    #
-    if test "$want_debug" = "yes"; then
-      { $as_echo "$as_me:${as_lineno-$LINENO}: checking if compiler accepts debug enabling options" >&5
-$as_echo_n "checking if compiler accepts debug enabling options... " >&6; }
-      tmp_options="$flags_dbg_yes"
-    fi
-    if test "$want_debug" = "no"; then
-      { $as_echo "$as_me:${as_lineno-$LINENO}: checking if compiler accepts debug disabling options" >&5
-$as_echo_n "checking if compiler accepts debug disabling options... " >&6; }
-      tmp_options="$flags_dbg_off"
-    fi
-    #
-    CPPFLAGS="$tmp_CPPFLAGS"
-    CFLAGS="$tmp_CFLAGS $tmp_options"
-    squeeze CPPFLAGS
-    squeeze CFLAGS
-
-    tmp_compiler_works="unknown"
-  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-
-
-int main (void)
-{
-
-      int i = 1;
-      return i;
-
- ;
- return 0;
-}
-
-_ACEOF
-if ac_fn_c_try_compile "$LINENO"; then :
-
-    tmp_compiler_works="yes"
-
-else
-
-    tmp_compiler_works="no"
-    echo " " >&6
-    sed 's/^/cc-fail: /' conftest.err >&6
-    echo " " >&6
-
-fi
-rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
-    if test "$tmp_compiler_works" = "yes"; then
-    cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-
-
-int main (void)
-{
-
-        int i = 1;
-        return i;
-
- ;
- return 0;
-}
-
-_ACEOF
-if ac_fn_c_try_link "$LINENO"; then :
-
-      tmp_compiler_works="yes"
-
-else
-
-      tmp_compiler_works="no"
-      echo " " >&6
-      sed 's/^/link-fail: /' conftest.err >&6
-      echo " " >&6
-
-fi
-rm -f core conftest.err conftest.$ac_objext \
-    conftest$ac_exeext conftest.$ac_ext
-  fi
-    if test "x$cross_compiling" != "xyes" &&
-    test "$tmp_compiler_works" = "yes"; then
-    if test "$cross_compiling" = yes; then :
-  { { $as_echo "$as_me:${as_lineno-$LINENO}: error: in \`$ac_pwd':" >&5
-$as_echo "$as_me: error: in \`$ac_pwd':" >&2;}
-as_fn_error $? "cannot run test program while cross compiling
-See \`config.log' for more details" "$LINENO" 5 ; }
-else
-  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-
-#       ifdef __STDC__
-#         include <stdlib.h>
-#       endif
-
-int main (void)
-{
-
-        int i = 0;
-        exit(i);
-
- ;
- return 0;
-}
-
-_ACEOF
-if ac_fn_c_try_run "$LINENO"; then :
-
-      tmp_compiler_works="yes"
-
-else
-
-      tmp_compiler_works="no"
-      echo " " >&6
-      echo "run-fail: test program exited with status $ac_status" >&6
-      echo " " >&6
-
-fi
-rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
-  conftest.$ac_objext conftest.beam conftest.$ac_ext
-fi
-
-  fi
-    if test "$tmp_compiler_works" = "yes"; then
-
-      { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
-$as_echo "yes" >&6; }
-      { $as_echo "$as_me:${as_lineno-$LINENO}: compiler options added: $tmp_options" >&5
-$as_echo "$as_me: compiler options added: $tmp_options" >&6;}
-
-  else
-
-      { $as_echo "$as_me:${as_lineno-$LINENO}: result: no" >&5
-$as_echo "no" >&6; }
-      { $as_echo "$as_me:${as_lineno-$LINENO}: WARNING: compiler options rejected: $tmp_options" >&5
-$as_echo "$as_me: WARNING: compiler options rejected: $tmp_options" >&2;}
-            CPPFLAGS="$tmp_save_CPPFLAGS"
-      CFLAGS="$tmp_save_CFLAGS"
-
-  fi
-
-    #
-  fi
 
 
         #
