diff -up curl-7.21.0/tests/data/Makefile.am.s390 curl-7.21.0/tests/data/Makefile.am
--- curl-7.21.0/tests/data/Makefile.am.s390	2010-07-12 12:11:12.000000000 +0200
+++ curl-7.21.0/tests/data/Makefile.am	2010-07-12 12:11:25.000000000 +0200
@@ -60,7 +60,7 @@ EXTRA_DIST = test1 test108 test117 test1
  test1072 test1073 test1074 test1075 test1076 test1077 test1078 test1079   \
  test1080 test1081 test1082 test1083 test1084 test1085 test633 test634     \
  test635 test636 test637 test558 test559 test1086 test1087 test1088        \
- test574 test575 test576 test577 test1113 test1114 \
+ test574 test576 test577 test1113 test1114 \
  test1089 test1090 test1091 test1092 test1093 test1094 test1095 test1096   \
  test1097 test560 test561 test1098 test1099 test562 test563 test1100       \
  test564 test1101 test1102 test1103 test1104 test299 test310 test311       \
