diff -up curl-7.26.0/tests/data/Makefile.am.ppc1319 curl-7.26.0/tests/data/Makefile.am
--- curl-7.26.0/tests/data/Makefile.am.ppc1319	2012-05-30 12:32:16.396720278 +0200
+++ curl-7.26.0/tests/data/Makefile.am	2012-05-30 12:32:40.475870630 +0200
@@ -81,7 +81,7 @@ test1208 test1209 test1210 test1211 \
 test1220 \
 test1300 test1301 test1302 test1303 test1304 test1305	\
 test1306 test1307 test1308 test1309 test1310 test1311 test1312 test1313 \
-test1314 test1315 test1316 test1317 test1318 test1319 test1320 test1321 \
+test1314 test1315 test1316 test1317 test1318 test1320 test1321 \
 test1322 test1323 test1324 test1325 test1326 test1327 test1328 test1329 \
 test1331 test1332 test1333 \
 test1400 test1401 test1402 test1403 test1404 test1405 test1406 test1407 \
diff -up curl-7.26.0/tests/data/Makefile.in.ppc1319 curl-7.26.0/tests/data/Makefile.in
--- curl-7.26.0/tests/data/Makefile.in.ppc1319	2012-05-30 12:32:24.891420536 +0200
+++ curl-7.26.0/tests/data/Makefile.in	2012-05-30 12:32:47.225632459 +0200
@@ -340,7 +340,7 @@ test1208 test1209 test1210 test1211 \
 test1220 \
 test1300 test1301 test1302 test1303 test1304 test1305	\
 test1306 test1307 test1308 test1309 test1310 test1311 test1312 test1313 \
-test1314 test1315 test1316 test1317 test1318 test1319 test1320 test1321 \
+test1314 test1315 test1316 test1317 test1318 test1320 test1321 \
 test1322 test1323 test1324 test1325 test1326 test1327 test1328 test1329 \
 test1331 test1332 test1333 \
 test1400 test1401 test1402 test1403 test1404 test1405 test1406 test1407 \
