diff --git a/lib/nss.c b/lib/nss.c
index 637663e..f439735 100644
--- a/lib/nss.c
+++ b/lib/nss.c
@@ -1047,7 +1047,7 @@ CURLcode Curl_nss_connect(struct connectdata *conn, int sockindex)
       }
       else {
         char *certpath = PR_smprintf("%s%s",
-                         NSS_VersionCheck("3.12.0") ? "sql:" : "",
+                         /*NSS_VersionCheck("3.12.0") ? "sql:" :*/ "",
                          certDir);
         rv = NSS_Initialize(certpath, "", "", "", NSS_INIT_READONLY);
         PR_smprintf_free(certpath);
