diff --git a/tests/runtests.pl b/tests/runtests.pl
index d4e9072..fb32c78 100755
--- a/tests/runtests.pl
+++ b/tests/runtests.pl
@@ -222,7 +222,7 @@ my $sshdverstr;  # for socks server, ssh daemon version string
 my $sshderror;   # for socks server, ssh daemon version error
 
 my $defserverlogslocktimeout = 20; # timeout to await server logs lock removal
-my $defpostcommanddelay = 0; # delay between command and postcheck sections
+my $defpostcommanddelay = 1; # delay between command and postcheck sections
 
 my $timestats;   # time stamping and stats generation
 my $fullstats;   # show time stats for every single test
