--- curl-7.16.1/libcurl.pc.in.privlibs	2007-01-02 13:14:21.000000000 +0100
+++ curl-7.16.1/libcurl.pc.in	2007-01-29 17:20:50.000000000 +0100
@@ -33,6 +33,6 @@
 URL: http://curl.haxx.se/
 Description: Library to transfer files with ftp, http, etc.
 Version: @VERSION@
-Libs: -L${libdir} -lcurl @LDFLAGS@
-Libs.private: @LIBS@
+Libs: -L${libdir} -lcurl
+Libs.private: @LDFLAGS@ @LIBS@
 Cflags: -I${includedir}
