diff -ruNp curl-7.19.3.orig/lib/nss.c curl-7.19.3/lib/nss.c
--- curl-7.19.3.orig/lib/nss.c	2009-01-07 15:12:01.000000000 +0100
+++ curl-7.19.3/lib/nss.c	2009-02-16 11:39:41.912075708 +0100
@@ -1140,7 +1140,7 @@ CURLcode Curl_nss_connect(struct connect
       n = strrchr(data->set.str[STRING_CERT], '/');
       if(n) {
         n++; /* skip last slash */
-        nickname = aprintf(nickname, "PEM Token #%d:%s", 1, n);
+        nickname = aprintf("PEM Token #%d:%s", 1, n);
         if(!nickname)
           return CURLE_OUT_OF_MEMORY;
 
@@ -1171,7 +1171,8 @@ CURLcode Curl_nss_connect(struct connect
 
     if(SSL_GetClientAuthDataHook(model,
                                  (SSLGetClientAuthData) SelectClientCert,
-                                 (void *)connssl) != SECSuccess) {
+                                 (void *)connssl->client_nickname) !=
+       SECSuccess) {
       curlerr = CURLE_SSL_CERTPROBLEM;
       goto error;
     }
