 lib/hostares.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/hostares.c b/lib/hostares.c
index 01b2142..ceda98e 100644
--- a/lib/hostares.c
+++ b/lib/hostares.c
@@ -380,7 +380,7 @@ Curl_addrinfo *Curl_getaddrinfo(struct connectdata *conn,
 
   switch(data->set.ip_version) {
   default:
-#if ARES_VERSION >= 0x010601
+#if 0
     family = PF_UNSPEC; /* supported by c-ares since 1.6.1, so for older
                            c-ares versions this just falls through and defaults
                            to PF_INET */
