diff -up curl-7.18.0/curl-config.in.curl-config curl-7.18.0/curl-config.in
--- curl-7.18.0/curl-config.in.curl-config	2008-01-28 21:39:08.000000000 +0100
+++ curl-7.18.0/curl-config.in	2008-01-28 21:46:36.000000000 +0100
@@ -57,6 +57,8 @@
     usage 1
 fi
 
+USE_SSL="@USE_SSLEAY@@USE_GNUTLS@@USE_NSS@"
+
 while test $# -gt 0; do
     case "$1" in
     # this deals with options in the style
@@ -81,9 +83,9 @@
 
     --feature|--features)
 	if test "@USE_SSLEAY@" = "1"; then
-          echo "SSL"
           NTLM=1 # OpenSSL implies NTLM
-        elif test -n "@SSL_ENABLED@"; then
+        fi
+        if test -n "$USE_SSL"; then
           echo "SSL"
         fi
 	if test "@KRB4_ENABLED@" = "1"; then
@@ -113,13 +115,13 @@
     --protocols)
 	if test "@CURL_DISABLE_HTTP@" != "1"; then
           echo "HTTP"
-	  if test "@SSL_ENABLED@" = "1"; then
+	  if test "$USE_SSL" = "1"; then
             echo "HTTPS"
           fi
         fi
 	if test "@CURL_DISABLE_FTP@" != "1"; then
           echo "FTP"
-	  if test "@SSL_ENABLED@" = "1"; then
+	  if test "$USE_SSL" = "1"; then
             echo "FTPS"
           fi
         fi
